File: ortools/java/com/google/ortools/modelbuilder/ModelBuilder.java
// Copyright 2010-2025 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package com.google.ortools.modelbuilder;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Main modeling class.
 *
 * <p>Proposes a factory to create all modeling objects understood by the SAT solver.
 */
public final class ModelBuilder {
  static class ModelBuilderException extends RuntimeException {
    public ModelBuilderException(String methodName, String msg) {
      // Call constructor of parent Exception
      super(methodName + ": " + msg);
    }
  }

  /** Exception thrown when parallel arrays have mismatched lengths. */
  public static class MismatchedArrayLengths extends ModelBuilderException {
    public MismatchedArrayLengths(String methodName, String array1Name, String array2Name) {
      super(methodName, array1Name + " and " + array2Name + " have mismatched lengths");
    }
  }

  /** Exception thrown when an array has a wrong length. */
  public static class WrongLength extends ModelBuilderException {
    public WrongLength(String methodName, String msg) {
      super(methodName, msg);
    }
  }

  /** Main constructor */
  public ModelBuilder() {
    helper = new ModelBuilderHelper();
    constantMap = new LinkedHashMap<>();
  }

  /** Returns a cloned model */
  public ModelBuilder getClone() {
    ModelBuilder clonedModel = new ModelBuilder();
    clonedModel.getHelper().overwriteModel(helper);
    clonedModel.constantMap.putAll(constantMap);
    return clonedModel;
  }

  // Integer variables.

  /** Creates a variable with domain [lb, ub]. */
  public Variable newVar(double lb, double ub, boolean isIntegral, String name) {
    return new Variable(helper, lb, ub, isIntegral, name);
  }

  /** Creates a continuous variable with domain [lb, ub]. */
  public Variable newNumVar(double lb, double ub, String name) {
    return new Variable(helper, lb, ub, false, name);
  }

  /** Creates an integer variable with domain [lb, ub]. */
  public Variable newIntVar(double lb, double ub, String name) {
    return new Variable(helper, lb, ub, true, name);
  }

  /** Creates a Boolean variable with the given name. */
  public Variable newBoolVar(String name) {
    return new Variable(helper, 0, 1, true, name);
  }

  /** Creates a constant variable. */
  public Variable newConstant(double value) {
    if (constantMap.containsKey(value)) {
      return new Variable(helper, constantMap.get(value));
    }
    Variable cste = new Variable(helper, value, value, false, ""); // bounds and name.
    constantMap.put(value, cste.getIndex());
    return cste;
  }

  /** Rebuilds a variable from its index. */
  public Variable varFromIndex(int index) {
    return new Variable(helper, index);
  }

  // Linear constraints.

  /** Adds {@code lb <= expr <= ub}. */
  public LinearConstraint addLinearConstraint(LinearArgument expr, double lb, double ub) {
    LinearConstraint lin = new LinearConstraint(helper);
    final LinearExpr e = expr.build();
    for (int i = 0; i < e.numElements(); ++i) {
      helper.addConstraintTerm(lin.getIndex(), e.getVariableIndex(i), e.getCoefficient(i));
    }
    double offset = e.getOffset();
    if (lb == Double.NEGATIVE_INFINITY || lb == Double.POSITIVE_INFINITY) {
      lin.setLowerBound(lb);
    } else {
      lin.setLowerBound(lb - offset);
    }
    if (ub == Double.NEGATIVE_INFINITY || ub == Double.POSITIVE_INFINITY) {
      lin.setUpperBound(ub);
    } else {
      lin.setUpperBound(ub - offset);
    }
    return lin;
  }

  /** Adds {@code expr == value}. */
  public LinearConstraint addEquality(LinearArgument expr, double value) {
    return addLinearConstraint(expr, value, value);
  }

  /** Adds {@code left == right}. */
  public LinearConstraint addEquality(LinearArgument left, LinearArgument right) {
    LinearExprBuilder difference = LinearExpr.newBuilder();
    difference.addTerm(left, 1);
    difference.addTerm(right, -1);
    return addLinearConstraint(difference, 0.0, 0.0);
  }

  /** Adds {@code expr <= value}. */
  public LinearConstraint addLessOrEqual(LinearArgument expr, double value) {
    return addLinearConstraint(expr, Double.NEGATIVE_INFINITY, value);
  }

  /** Adds {@code left <= right}. */
  public LinearConstraint addLessOrEqual(LinearArgument left, LinearArgument right) {
    LinearExprBuilder difference = LinearExpr.newBuilder();
    difference.addTerm(left, 1);
    difference.addTerm(right, -1);
    return addLinearConstraint(difference, Double.NEGATIVE_INFINITY, 0.0);
  }

  /** Adds {@code expr >= value}. */
  public LinearConstraint addGreaterOrEqual(LinearArgument expr, double value) {
    return addLinearConstraint(expr, value, Double.POSITIVE_INFINITY);
  }

  /** Adds {@code left >= right}. */
  public LinearConstraint addGreaterOrEqual(LinearArgument left, LinearArgument right) {
    LinearExprBuilder difference = LinearExpr.newBuilder();
    difference.addTerm(left, 1);
    difference.addTerm(right, -1);
    return addLinearConstraint(difference, 0.0, Double.POSITIVE_INFINITY);
  }

  /** Rebuilds a linear constraint from its index. */
  public LinearConstraint constraintFromIndex(int index) {
    return new LinearConstraint(helper, index);
  }

  // Enforced Linear constraints.

  /** Adds {@code ivar == iValue => lb <= expr <= ub}. */
  public EnforcedLinearConstraint addEnforcedLinearConstraint(
      LinearArgument expr, double lb, double ub, Variable iVar, boolean iValue) {
    EnforcedLinearConstraint lin = new EnforcedLinearConstraint(helper);
    lin.setIndicatorVariable(iVar);
    lin.setIndicatorValue(iValue);
    final LinearExpr e = expr.build();
    for (int i = 0; i < e.numElements(); ++i) {
      helper.addEnforcedConstraintTerm(lin.getIndex(), e.getVariableIndex(i), e.getCoefficient(i));
    }
    double offset = e.getOffset();
    if (lb == Double.NEGATIVE_INFINITY || lb == Double.POSITIVE_INFINITY) {
      lin.setLowerBound(lb);
    } else {
      lin.setLowerBound(lb - offset);
    }
    if (ub == Double.NEGATIVE_INFINITY || ub == Double.POSITIVE_INFINITY) {
      lin.setUpperBound(ub);
    } else {
      lin.setUpperBound(ub - offset);
    }
    return lin;
  }

  /** Adds {@code ivar == iValue => expr == value}. */
  public EnforcedLinearConstraint addEnforcedEquality(
      LinearArgument expr, double value, Variable iVar, boolean iValue) {
    return addEnforcedLinearConstraint(expr, value, value, iVar, iValue);
  }

  /** Adds {@code ivar == iValue => left == right}. */
  public EnforcedLinearConstraint addEnforcedEquality(
      LinearArgument left, LinearArgument right, Variable iVar, boolean iValue) {
    LinearExprBuilder difference = LinearExpr.newBuilder();
    difference.addTerm(left, 1);
    difference.addTerm(right, -1);
    return addEnforcedLinearConstraint(difference, 0.0, 0.0, iVar, iValue);
  }

  /** Adds {@code ivar == iValue => expr <= value}. */
  public EnforcedLinearConstraint addEnforcedLessOrEqual(
      LinearArgument expr, double value, Variable iVar, boolean iValue) {
    return addEnforcedLinearConstraint(expr, Double.NEGATIVE_INFINITY, value, iVar, iValue);
  }

  /** Adds {@code ivar == iValue => left <= right}. */
  public EnforcedLinearConstraint addEnforcedLessOrEqual(
      LinearArgument left, LinearArgument right, Variable iVar, boolean iValue) {
    LinearExprBuilder difference = LinearExpr.newBuilder();
    difference.addTerm(left, 1);
    difference.addTerm(right, -1);
    return addEnforcedLinearConstraint(difference, Double.NEGATIVE_INFINITY, 0.0, iVar, iValue);
  }

  /** Adds {@code ivar == iValue => expr >= value}. */
  public EnforcedLinearConstraint addEnforcedGreaterOrEqual(
      LinearArgument expr, double value, Variable iVar, boolean iValue) {
    return addEnforcedLinearConstraint(expr, value, Double.POSITIVE_INFINITY, iVar, iValue);
  }

  /** Adds {@code ivar == iValue => left >= right}. */
  public EnforcedLinearConstraint addEnforcedGreaterOrEqual(
      LinearArgument left, LinearArgument right, Variable iVar, boolean iValue) {
    LinearExprBuilder difference = LinearExpr.newBuilder();
    difference.addTerm(left, 1);
    difference.addTerm(right, -1);
    return addEnforcedLinearConstraint(difference, 0.0, Double.POSITIVE_INFINITY, iVar, iValue);
  }

  /** Rebuilds a linear constraint from its index. */
  public EnforcedLinearConstraint enforcedConstraintFromIndex(int index) {
    return new EnforcedLinearConstraint(helper, index);
  }

  /** Minimize expression */
  public void minimize(LinearArgument obj) {
    optimize(obj, false);
  }

  /** Minimize expression */
  public void maximize(LinearArgument obj) {
    optimize(obj, true);
  }

  /** Sets the objective expression. */
  public void optimize(LinearArgument obj, boolean maximize) {
    helper.clearObjective();
    LinearExpr e = obj.build();
    LinkedHashMap<Integer, Double> coeffMap = new LinkedHashMap<>();
    for (int i = 0; i < e.numElements(); ++i) {
      coeffMap.merge(e.getVariableIndex(i), e.getCoefficient(i), Double::sum);
    }
    for (Map.Entry<Integer, Double> entry : coeffMap.entrySet()) {
      if (entry.getValue() != 0) {
        helper.setVarObjectiveCoefficient(entry.getKey(), entry.getValue());
      }
    }
    helper.setObjectiveOffset(e.getOffset());
    helper.setMaximize(maximize);
  }

  /** returns the objective offset. */
  double getObjectiveOffset() {
    return helper.getObjectiveOffset();
  }

  /** Sets the objective offset. */
  void setObjectiveOffset(double offset) {
    helper.setObjectiveOffset(offset);
  }

  /** Clears all hints from the solver. */
  void clearHints() {
    helper.clearHints();
  }

  /**
   * Adds var == value as a hint to the model. Note that variables must not appear more than once in
   * the list of hints.
   */
  void addHint(Variable v, double value) {
    helper.addHint(v.getIndex(), value);
  }

  // Model getters, import, export.

  /** Returns the number of variables in the model. */
  public int numVariables() {
    return helper.numVariables();
  }

  /** Returns the number of constraints in the model. */
  public int numConstraints() {
    return helper.numConstraints();
  }

  /** Returns the name of the model. */
  public String getName() {
    return helper.getName();
  }

  /** Sets the name of the model. */
  public void setName(String name) {
    helper.setName(name);
  }

  /**
   * Write the model as a protocol buffer to 'file'.
   *
   * @param file file to write the model to. If the filename ends with 'txt', the model will be
   *     written as a text file, otherwise, the binary format will be used.
   * @return true if the model was correctly written.
   */
  public boolean exportToFile(String file) {
    return helper.writeModelToProtoFile(file);
  }

  /**
   * import the model from protocol buffer 'file'.
   *
   * @param file file to read the model from.
   * @return true if the model was correctly loaded.
   */
  public boolean importFromFile(String file) {
    return helper.readModelFromProtoFile(file);
  }

  public String exportToMpsString(boolean obfuscate) {
    return helper.exportToMpsString(obfuscate);
  }

  public String exportToLpString(boolean obfuscate) {
    return helper.exportToLpString(obfuscate);
  }

  public boolean writeToMpsFile(String filename, boolean obfuscate) {
    return helper.writeToMpsFile(filename, obfuscate);
  }

  public boolean importFromMpsString(String mpsString) {
    return helper.importFromMpsString(mpsString);
  }

  public boolean importFromMpsFile(String mpsFile) {
    return helper.importFromMpsString(mpsFile);
  }

  public boolean importFromLpString(String lpString) {
    return helper.importFromLpString(lpString);
  }

  public boolean importFromLpFile(String lpFile) {
    return helper.importFromMpsString(lpFile);
  }

  // Getters.
  /** Returns the model builder helper. */
  public ModelBuilderHelper getHelper() {
    return helper;
  }

  private final ModelBuilderHelper helper;
  private final Map<Double, Integer> constantMap;
}

File: ortools/linear_solver/csharp/ModelBuilder.cs
// Copyright 2010-2025 Google LLC
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

namespace Google.OrTools.ModelBuilder
{

using Google.OrTools.Util;
using System;
using System.Collections;
using System.Collections.Generic;
using System.Linq;
using System.Runtime.CompilerServices;
using Google.Protobuf.Collections;

/// <summary>
///  Main modeling class.
/// </summary>
///
/// Proposes a factory to create all modeling objects understood by the Solver.
public class Model
{
    /// <summary>
    /// Main constructor.
    /// </summary>
    public Model()
    {
        helper_ = new ModelBuilderHelper();
        constantMap_ = new Dictionary<double, int>();
        tmp_var_value_map_ = new SortedDictionary<int, double>();
        tmp_terms_ = new Queue<Term>();
    }

    /// <summary>
    /// Returns a cloned model.
    /// </summary>
    /// <returns>A deep copy of the model.</returns>
    public Model Clone()
    {
        Model clonedModel = new Model();
        clonedModel.Helper.OverwriteModel(Helper);
        foreach (KeyValuePair<double, int> entry in constantMap_)
        {
            clonedModel.constantMap_[entry.Key] = entry.Value;
        }
        return clonedModel;
    }

    // Integer variables.

    /// <summary>
    /// reates a variable with domain [lb, ub].
    /// </summary>
    /// <param name="lb">The lower bound of the variable</param>
    /// <param name="ub">The upper bound of the variable</param>
    /// <param name="isIntegral">Indicates if the variable is restricted to take only integral values</param>
    /// <param name="name">The name of the variable</param>
    /// <returns>The new variable</returns>
    public Variable NewVar(double lb, double ub, bool isIntegral, String name)
    {
        return new Variable(helper_, lb, ub, isIntegral, name);
    }

    /// <summary>
    /// Creates a continuous variable with domain [lb, ub].
    /// </summary>
    /// <param name="lb">The lower bound of the variable</param>
    /// <param name="ub">The upper bound of the variable</param>
    /// <param name="name">The name of the variable</param>
    /// <returns>The new continuous variable</returns>
    public Variable NewNumVar(double lb, double ub, String name)
    {
        return new Variable(helper_, lb, ub, false, name);
    }

    /// <summary>
    /// Creates an integer variable with domain [lb, ub].
    /// </summary>
    /// <param name="lb">The lower bound of the variable</param>
    /// <param name="ub">The upper bound of the variable</param>
    /// <param name="name">The name of the variable</param>
    /// <returns>The new integer variable</returns>
    public Variable NewIntVar(double lb, double ub, String name)
    {
        return new Variable(helper_, lb, ub, true, name);
    }

    /// <summary>
    /// Creates a bool variable with the given name.
    /// </summary>
    /// <param name="name">The name of the variable</param>
    /// <returns>The new Boolean variable</returns>
    public Variable NewBoolVar(String name)
    {
        return new Variable(helper_, 0, 1, true, name);
    }

    /// <summary>
    /// Creates a constant variable.
    /// </summary>
    /// <param name="value">the value of the constant variable</param>
    /// <returns>A new variable with a fixed value</returns>
    public Variable NewConstant(double value)
    {
        if (constantMap_.TryGetValue(value, out int index))
        {
            return new Variable(helper_, index);
        }
        Variable cste = new Variable(helper_, value, value, false, ""); // bounds and name.
        constantMap_.Add(value, cste.Index);
        return cste;
    }

    /// Rebuilds a variable from its index.
    public Variable VarFromIndex(int index)
    {
        return new Variable(helper_, index);
    }

    /// <summary>
    /// Adds a Linear constraint to the model.
    /// </summary>
    /// <param name="lin">A bounded linear expression</param>
    /// <returns>A linear expression</returns>
    /// <exception cref="ArgumentException">Throw when the constraint is not supported by the linear solver</exception>
    public LinearConstraint Add(BoundedLinearExpression lin)
    {
        switch (lin.CtType)
        {
        case BoundedLinearExpression.Type.BoundExpression: {
            return AddLinearConstraint(lin.Left, lin.Lb, lin.Ub);
        }
        case BoundedLinearExpression.Type.VarEqVar: {
            return AddLinearConstraint(lin.Left - lin.Right, 0, 0);
        }
        case BoundedLinearExpression.Type.VarEqCst: {
            return AddLinearConstraint(lin.Left, lin.Lb, lin.Lb);
        }
        default: {
            throw new ArgumentException("Cannot use '" + lin.ToString() + "' as a linear constraint.");
        }
        }
    }

    /// <summary>
    /// Adds the constraint expr in [lb, ub].
    /// </summary>
    /// <param name="expr">The constrained expression</param>
    /// <param name="lb">the constrained lower bound of the expression</param>
    /// <param name="ub">the constrained upper bound of the expression</param>
    /// <returns>the linear constraint</returns>
    public LinearConstraint AddLinearConstraint(LinearExpr expr, double lb, double ub)
    {
        LinearConstraint lin = new LinearConstraint(helper_);
        var dict = tmp_var_value_map_;
        dict.Clear();
        double offset = LinearExpr.GetVarValueMap(expr, dict, tmp_terms_);
        foreach (KeyValuePair<int, double> term in dict)
        {
            helper_.AddConstraintTerm(lin.Index, term.Key, term.Value);
        }
        if (lb == Double.NegativeInfinity || lb == Double.PositiveInfinity)
        {
            lin.LowerBound = lb;
        }
        else
        {
            lin.LowerBound = lb - offset;
        }
        if (ub == Double.NegativeInfinity || ub == Double.PositiveInfinity)
        {
            lin.UpperBound = ub;
        }
        else
        {
            lin.UpperBound = ub - offset;
        }
        return lin;
    }

    /// Rebuilds a linear constraint from its index.
    public LinearConstraint ConstraintFromIndex(int index)
    {
        return new LinearConstraint(helper_, index);
    }

    /// <summary>
    /// Adds an enforced Linear constraint to the model.
    /// </summary>
    /// <param name="lin">A bounded linear expression</param>
    /// <param name="iVar">The indicator variable of the constraint.</param>
    /// <param name="iValue">The indicator value of the constraint.</param>
    /// <returns>A linear expression</returns>
    /// <exception cref="ArgumentException">Throw when the constraint is not supported by the linear solver</exception>
    public EnforcedLinearConstraint AddEnforced(BoundedLinearExpression lin, Variable iVar, bool iValue)
    {
        switch (lin.CtType)
        {
        case BoundedLinearExpression.Type.BoundExpression: {
            return AddEnforcedLinearConstraint(lin.Left, lin.Lb, lin.Ub, iVar, iValue);
        }
        case BoundedLinearExpression.Type.VarEqVar: {
            return AddEnforcedLinearConstraint(lin.Left - lin.Right, 0, 0, iVar, iValue);
        }
        case BoundedLinearExpression.Type.VarEqCst: {
            return AddEnforcedLinearConstraint(lin.Left, lin.Lb, lin.Lb, iVar, iValue);
        }
        default: {
            throw new ArgumentException("Cannot use '" + lin.ToString() + "' as a linear constraint.");
        }
        }
    }

    /// <summary>
    /// Adds the constraint iVar == iValue => expr in [lb, ub].
    /// </summary>
    /// <param name="expr">The constrained expression</param>
    /// <param name="lb">the lower bound of the constraint</param>
    /// <param name="ub">the upper bound of the constraint</param>
    /// <param name="iVar">the indicator variable of the constraint</param>
    /// <param name="iValue">the indicator value of the constraint</param>
    /// <returns>the enforced linear constraint</returns>
    public EnforcedLinearConstraint AddEnforcedLinearConstraint(LinearExpr expr, double lb, double ub, Variable iVar,
                                                                bool iValue)
    {
        EnforcedLinearConstraint lin = new EnforcedLinearConstraint(helper_);
        lin.IndicatorVariable = iVar;
        lin.IndicatorValue = iValue;
        var dict = tmp_var_value_map_;
        dict.Clear();
        double offset = LinearExpr.GetVarValueMap(expr, dict, tmp_terms_);
        foreach (KeyValuePair<int, double> term in dict)
        {
            helper_.AddEnforcedConstraintTerm(lin.Index, term.Key, term.Value);
        }
        if (lb == Double.NegativeInfinity || lb == Double.PositiveInfinity)
        {
            lin.LowerBound = lb;
        }
        else
        {
            lin.LowerBound = lb - offset;
        }
        if (ub == Double.NegativeInfinity || ub == Double.PositiveInfinity)
        {
            lin.UpperBound = ub;
        }
        else
        {
            lin.UpperBound = ub - offset;
        }
        return lin;
    }

    /// Rebuilds a linear constraint from its index.
    public EnforcedLinearConstraint EnforcedConstraintFromIndex(int index)
    {
        return new EnforcedLinearConstraint(helper_, index);
    }

    // Objective.

    /// <summary>
    /// Minimize expression.
    /// </summary>
    /// <param name="obj">the linear expression to minimize</param>
    public void Minimize(LinearExpr obj)
    {
        Optimize(obj, false);
    }

    /// <summary>
    /// Maximize expression.
    /// </summary>
    /// <param name="obj">the linear expression to maximize</param>
    public void Maximize(LinearExpr obj)
    {
        Optimize(obj, true);
    }

    /// <summary>
    /// Sets the objective expression.
    /// </summary>
    /// <param name="obj">the linear expression to optimize</param>
    /// <param name="maximize">the direction of the optimization</param>
    public void Optimize(LinearExpr obj, bool maximize)
    {
        helper_.ClearObjective();
        var dict = tmp_var_value_map_;
        dict.Clear();
        double offset = LinearExpr.GetVarValueMap(obj, dict, tmp_terms_);
        foreach (KeyValuePair<int, double> term in dict)
        {
            if (term.Value != 0.0)
            {
                helper_.SetVarObjectiveCoefficient(term.Key, term.Value);
            }
        }
        helper_.SetObjectiveOffset(offset);
        helper_.SetMaximize(maximize);
    }

    /// <summary>
    /// The offset of the objective.
    /// </summary>
    public double ObjectiveOffset
    {
        get {
            return helper_.ObjectiveOffset();
        }
        set {
            helper_.SetObjectiveOffset(value);
        }
    }

    /// <summary>
    /// Remove all hints from the model.
    /// </summary>
    public void ClearHints()
    {
        helper_.ClearHints();
    }

    /// <summary>
    /// Adds var == value as a hint to the model.  Note that variables must not appear more than once in the list of
    /// hints.
    /// </summary>
    public void AddHint(Variable var, double value)
    {
        helper_.AddHint(var.Index, value);
    }

    /// <summary>
    /// Returns the number of variables in the model.
    /// </summary>
    public int VariablesCount()
    {
        return helper_.VariablesCount();
    }

    /// <summary>
    /// Returns the number of constraints in the model.
    /// </summary>
    public int ConstraintsCount()
    {
        return helper_.ConstraintsCount();
    }

    /// <summary>
    /// The name of the model.
    /// </summary>
    public String Name
    {
        get {
            return helper_.Name();
        }
        set {
            helper_.SetName(value);
        }
    }

    /// <summary>
    /// Write the model as a protocol buffer to 'file'.
    /// </summary>
    /// @param file file to write the model to. If the filename ends with 'txt', the model will be
    ///    written as a text file, otherwise, the binary format will be used.
    ///@return true if the model was correctly written.
    public bool ExportToFile(String file)
    {
        return helper_.WriteModelToProtoFile(file);
    }

    /// <summary>
    /// load the model as a protocol buffer from 'file'.
    /// </summary>
    /// @param file file to read the model from.
    ///@return true if the model was correctly loaded.
    public bool ImportFromFile(String file)
    {
        return helper_.ReadModelFromProtoFile(file);
    }

    public String ExportToMpsString(bool obfuscate)
    {
        return helper_.ExportToMpsString(obfuscate);
    }

    public String ExportToLpString(bool obfuscate)
    {
        return helper_.ExportToLpString(obfuscate);
    }

    public bool WriteToMpsFile(String filename, bool obfuscate)
    {
        return helper_.WriteToMpsFile(filename, obfuscate);
    }

    public bool ImportFromMpsString(String mpsString)
    {
        return helper_.ImportFromMpsString(mpsString);
    }

    public bool ImportFromMpsFile(String mpsFile)
    {
        return helper_.ImportFromMpsString(mpsFile);
    }

    public bool ImportFromLpString(String lpString)
    {
        return helper_.ImportFromLpString(lpString);
    }

    public bool ImportFromLpFile(String lpFile)
    {
        return helper_.ImportFromMpsString(lpFile);
    }

    /// <summary>
    /// The model builder helper.
    /// </summary>
    public ModelBuilderHelper Helper
    {
        get {
            return helper_;
        }
    }

    private ModelBuilderHelper helper_;
    private Dictionary<double, int> constantMap_;

    // Used to process linear expressions.
    private SortedDictionary<int, double> tmp_var_value_map_;
    private Queue<Term> tmp_terms_;
}

} // namespace Google.OrTools.ModelBuilder
